core/client/fs/src/main/java/alluxio/client/metrics/MetricsHeartbeatContext.java

```diff
@@ -0,0 +1,241 @@
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.client.metrics;
+
+import alluxio.ClientContext;
+import alluxio.conf.AlluxioConfiguration;
+import alluxio.conf.PropertyKey;
+import alluxio.master.MasterClientContext;
+import alluxio.master.MasterInquireClient;
+import alluxio.util.ThreadFactoryUtils;
+import alluxio.util.network.NetworkAddressUtils;
+
+import com.google.common.base.Preconditions;
+import net.jcip.annotations.ThreadSafe;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * A class used to track metrics heartbeats to a master.
+ *
+ * The class should be instantiated when a new FileSystemContext is created with a
+ * configuration that points to a given master. As new FileSystemContexts are created, if they
+ * connect to the same master RPC address, then they can simply be added to this context so
+ * that their information is included in the metrics heartbeat.
+ *
+ * For each separate master RPC address, a new instance of this class is created. As
+ * FileSystemContexts are closed, they remove their application Ids from the internal metrics
+ * heartbeat. When a context reaches 0 applicationIds it will automatically close and remove
+ * itself from the internal MASTER_METRICS_HEARTBEAT map.
+ *
+ * When the final FileSystemContext closes and removes its heartbeat from metrics it will also
+ * shutdown and close the executor service until a new FileSystemContext is created.
+ */
+@ThreadSafe
+public class MetricsHeartbeatContext {
+  private static final Logger LOG = LoggerFactory.getLogger(MetricsHeartbeatContext.class);
+
+  /** A map from master RPC address to heartbeat context instances. */
+  private static final ConcurrentHashMap<String, MetricsHeartbeatContext>
+      MASTER_METRICS_HEARTBEAT = new ConcurrentHashMap<>();
+
+  /** A value that tracks whether or not we've registered the shutdown hook. */
+  private static boolean sAddedShudownHook = false;
+
+  /** The service which executes metrics heartbeats. */
+  private static ScheduledExecutorService sExecutorService;
+
+  private final String mMasterAddr;
+  private final MetricsMasterClient mMetricsMasterClient;
+  private final ClientMasterSync mClientMasterSync;
+  private ScheduledFuture<?> mMetricsMasterHeartbeatTask;
+  private final AtomicInteger mCtxCount;
```

Comment:
Does this need to be an `AtomicInteger`? All accesses are synchronized (aside from the constructor)?

Reply:
In a previous implementation this wasn't the case, but I could change this to a normal integer.